src/workerd/api/messagechannel.c++

```diff
@@ -0,0 +1,317 @@
+      state = Started{};
+      for (auto& item: list) {
+        ioContext.addTask(sendToRpc(capnp::clone(*item)));
+        // Local delivery
+        deliver(js, kj::mv(item), addRef());
+      }
+    }
+    KJ_CASE_ONEOF(_, Started) {
+      // Nothing to do in this case. We are already started!
+    }
+    KJ_CASE_ONEOF(_, Closed) {
+      // Nothing to do in this case. Can't start after closing.
+    }
+  }
+}
+
+kj::Maybe<jsg::JsValue> MessagePort::getOnMessage(jsg::Lock& js) {
+  return onmessageValue.map(
+      [&](jsg::JsRef<jsg::JsValue>& ref) -> jsg::JsValue { return ref.getHandle(js); });
+}
+
+void MessagePort::setOnMessage(jsg::Lock& js, jsg::JsValue value) {
+  if (!value.isObject() && !value.isFunction()) {
+    onmessageValue = kj::none;
+    // If we have no handlers and no onmessage ...
+    if (getHandlerCount("message"_kj) == 0 && onmessageValue == kj::none) {
+      // ...Put the port back into a pending state where messages
+      // will be enqueued until another listener is attached.
+      state = Pending();
+    }
+  } else {
+    onmessageValue = jsg::JsRef<jsg::JsValue>(js, value);
+    start(js);
+  }
+}
+
+namespace {
+// The jsrpc handler that receives messages posted from the remote and
+// delivers them to the local port.
+class JsMessagePortImpl final: public rpc::JsMessagePort::Server {
+ public:
+  JsMessagePortImpl(IoContext& ctx, jsg::Ref<MessagePort> port)
+      : port(kj::mv(port)),
+        weakIoContext(ctx.getWeakRef()) {}
+  ~JsMessagePortImpl() {
+    port->close();
+  }
+
+  kj::Promise<void> call(CallContext context) override {
+    IoContext& ctx = JSG_REQUIRE_NONNULL(weakIoContext->tryGet(), Error,
+        "The destination object for this message port no longer exists.");
+
+    KJ_IF_SOME(other, port->getOther()) {
+      // We can only dispatch messages under the isolate lock, so acquire
+      // that here then deliver...
+      auto lock = co_await ctx.getWorker()->takeAsyncLockWithoutRequest(nullptr);
+      ctx.getWorker()->runInLockScope(lock, [&](auto& lock) {
+        JSG_WITHIN_CONTEXT_SCOPE(lock, lock.getContext(), [&](auto& js) {
+          auto params = context.getParams();
+          other.deliverMessage(js, params.getData());
+        });
+      });
+    }
+
+    context.initResults();
+  }
+
+ private:
+  jsg::Ref<MessagePort> port;
+  kj::Own<IoContext::WeakRef> weakIoContext;
+};
+}  // namespace
+
+void MessagePort::serialize(jsg::Lock& js, jsg::Serializer& serializer) {
+  auto& handler = JSG_REQUIRE_NONNULL(serializer.getExternalHandler(), DOMDataCloneError,
+      "MessagePort can only be serialized for RPC.");
+  auto externalHandler = dynamic_cast<RpcSerializerExternalHander*>(&handler);
+  JSG_REQUIRE(
+      externalHandler != nullptr, DOMDataCloneError, "MessagePort can only be serialized for RPC.");
+
+  // Don't send a MessagePort that has been closed already.
+  JSG_REQUIRE(!state.is<Closed>(), DOMDataCloneError, "MessagePort is closed");
```

Comment:
This should be DOMException not DOMDataCloneError

> If [this](https://webidl.spec.whatwg.org/#this)'s [closed flag](https://html.spec.whatwg.org/multipage/web-messaging.html#concept-broadcastchannel-closed) is true, then throw an ["InvalidStateError"](https://webidl.spec.whatwg.org/#invalidstateerror) [DOMException](https://webidl.spec.whatwg.org/#dfn-DOMException).

Reply:
`DOMDataCloneError` *is* a `DOMException`